### PR TITLE
Subsystem for Power Logging

### PR DESCRIPTION
**Pull request by 1toldyou**
still working on this
especially not sure I'm imported the correct hardware or not

### CODE REVIEW

**Comment by 1toldyou** [comment]
having some difficulties on initialize the PDP (unable to find the CAN ID)
since the new robot will have different brand of PDP and wiring, so I'll wait till the new robot finished to continue

File: src/main/java/frc/team2412/robot/Hardware.java [hunk 1]
```diff
@@ -78,6 +80,15 @@
...
         public static final int LEFT_INTAKE_COLORSENSOR_PORT = 1, RIGHT_INTAKE_COLORSENSOR_PORT = 2,
                 CENTER_INTAKE_COLORSENSOR_PORT = 3;
+
+        /*
+         * According to
+         * https://docs.wpilib.org/en/stable/docs/software/can-devices/power-distribution-module.html
+         * The CAN ID for CTRE should be 0
+         * Or 1 for Rev (we're using Rev for the new robot)
+         */
+        public static final int PDP_CAN_ID = 9;
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/Hardware.java:90]: This isn't 0 or 1.  Do we use a non-default CAN ID for the power distribution panel?  If so, add a comment saying that so someone doesn't "fix" this and break it.

File: src/main/java/frc/team2412/robot/Hardware.java [hunk 2]
```diff
@@ -144,5 +158,9 @@ public Hardware() {
         if (GOAL_VIS_ENABLED) {
             limelight = new PhotonCamera(LIMELIGHT);
         }
+        if (MONITOR_ENABLED) {
+            powerDistributionPanel = new PowerDistribution(PDP_CAN_ID, PDP_MODULE_TYPE);
+            // powerDistributionPanel = new PowerDistribution();
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/Hardware.java:163]: Delete unused code

**Review by jamesdooley4 — APPROVED** [review]

File: src/main/java/frc/team2412/robot/subsystem/MonitoringSubsystem.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import edu.wpi.first.networktables.NetworkTableEntry;
+import edu.wpi.first.wpilibj.PowerDistribution;
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+import edu.wpi.first.wpilibj.shuffleboard.ShuffleboardTab;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class MonitoringSubsystem extends SubsystemBase {
+    ShuffleboardTab tab = Shuffleboard.getTab("Power Monitoring");
+    NetworkTableEntry totalPower;
+    PowerDistribution powerDistributionPanel;
```
Comment (TAKBS2412) [inline src/main/java/frc/team2412/robot/subsystem/MonitoringSubsystem.java:12]: These three variables should start as `private` - we can change later if needed.

**Review by TAKBS2412 — APPROVED** [review]
To find the CAN ID for the CTRE power distribution panel, there is the Phoenix Tuner: [https://docs.ctre-phoenix.com/en/stable/ch03_PrimerPhoenixSoft.html](https://docs.ctre-phoenix.com/en/stable/ch03_PrimerPhoenixSoft.html).

For the Rev Robotics power distribution hub, there is the REV hardware client: [https://docs.revrobotics.com/rev-11-1850/rev-hardware-client/getting-started-with-the-rev-hardware-client](https://docs.revrobotics.com/rev-11-1850/rev-hardware-client/getting-started-with-the-rev-hardware-client).

**Review by bobtseattle — COMMENTED** [review]
Sumedh has discovered there may be a plugin for Shuffleboard that displays this more easily than coding a subsystem:
- https://github.com/wpilibsuite/allwpilib/blob/main/wpilibj/src/main/java/edu/wpi/first/wpilibj/shuffleboard/BuiltInWidgets.java
Would have been nice of them to mention somewhere in the docs.

File: src/main/java/frc/team2412/robot/subsystem/MonitoringSubsystem.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import edu.wpi.first.wpilibj.shuffleboard.ShuffleboardTab;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class MonitoringSubsystem extends SubsystemBase {
+    private final ShuffleboardTab tab = Shuffleboard.getTab("Power Monitoring");
+    private final NetworkTableEntry totalPower, totalCurrent, currentTemperature,
+            channel0current, channel1current, channel2current, channel3current, channel4current, channel5current,
+            channel6current, channel7current,
+            channel8current, channel9current, channel10current, channel11current, channel12current, channel13current,
+            channel14current, channel15current;
```
Comment (jamesdooley4) [inline src/main/java/frc/team2412/robot/subsystem/MonitoringSubsystem.java:15]: Not sure this is necessary anymore, as the driver station logs include the current values for all the PDM channels already.

**Review by jamesdooley4 — CHANGES_REQUESTED** [review]

**Comment by TAKBS2412** [comment]
It seems that the data logged here is also logged automatically to the driver station. If that is the case, should we just close this PR?

**Pull request closed without merging**